subprojects/model-core/src/main/java/org/gradle/model/internal/type/ClassTypeWrapper.java

```diff
@@ -109,4 +110,21 @@ private String tryToGetRepresentation(boolean full) {
         } while (clazz != null);
         return classChain;
     }
+
+    private Class<?> effectiveClassOf(Class<?> clazz) {
+        Class<?> enclosingClass = clazz.getEnclosingClass();
+
+        // Detect anonymous static classes of enum constants with class body
+        // See https://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html#d5e12300
+        // And https://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html#jls-8.9.1
+        // "The optional class body of an enum constant implicitly defines an anonymous class declaration
+        //  that extends the immediately enclosing enum type."
+        if (enclosingClass != null && enclosingClass.isEnum()
```

Comment:
:o: minor thing: I'd split the line before `&& enclosingClass.isEnum()` for consistency with the other operands.